#! /bin/sh /usr/share/dpatch/dpatch-run
## 16_lirc-gpio.dpatch by Mario Limonciello <superm1@ubuntu.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Switch all GPIO based drivers to use dev/input interface

@DPATCH@
diff -urNad lirc-0.8.3~/doc/lirc.hwdb lirc-0.8.3/doc/lirc.hwdb
--- lirc-0.8.3~/doc/lirc.hwdb	2008-05-10 16:29:28.000000000 -0500
+++ lirc-0.8.3/doc/lirc.hwdb	2008-05-10 16:43:26.000000000 -0500
@@ -57,46 +57,46 @@
 
 [TV card]
 Adaptec AVC-2410;adaptec;lirc_dev lirc_i2c;hw_default;adaptec/lircd.conf.AVC-2410;
-Askey Magic TView CPH03x (card=1);cph03x;lirc_dev lirc_gpio;hw_default;cph03x/lircd.conf.cph03x;
-Askey/Typhoon/Anubis Magic TView CPH051/061 (bt878) (card=24);cph06x;lirc_dev lirc_gpio;hw_default;cph03x/lircd.conf.cph03x;
+Askey Magic TView CPH03x (card=1);cph03x;devinput;hw_default;cph03x/lircd.conf.cph03x;
+Askey/Typhoon/Anubis Magic TView CPH051/061 (bt878) (card=24);cph06x;devinput;hw_default;cph03x/lircd.conf.cph03x;
 Asus TV-Box;tvbox;lirc_dev lirc_i2c;hw_default;asus/lircd.conf.asus;
-AverMedia TV card (TVCapture, TVPhone) (card=6);avermedia;lirc_dev lirc_gpio;hw_default;avermedia/lircd.conf.avermedia;
-AverMedia TV card (TVCapture98, TVPhone98) (card=13/41);avermedia98;lirc_dev lirc_gpio;hw_default;avermedia/lircd.conf.avermedia98;
-AverMedia TV card (VDOMATE) (use card=13);avermedia_vdomate;lirc_dev lirc_gpio;hw_default;avermedia/lircd.conf.vdomate;
-BestBuy Easy TV (BT848) (card=55);bestbuy;lirc_dev lirc_gpio;hw_default;bestbuy/lircd.conf.bestbuy;
-BestBuy Easy TV (BT878) (card=62);bestbuy2;lirc_dev lirc_gpio;hw_default;bestbuy/lircd.conf.bestbuy2;
-Chronos Video Shuttle II (card=35);chronos;lirc_dev lirc_gpio;hw_default;chronos/lircd.conf.chronos;
+AverMedia TV card (TVCapture, TVPhone) (card=6);avermedia;devinput;hw_default;avermedia/lircd.conf.avermedia;
+AverMedia TV card (TVCapture98, TVPhone98) (card=13/41);avermedia98;devinput;hw_default;avermedia/lircd.conf.avermedia98;
+AverMedia TV card (VDOMATE) (use card=13);avermedia_vdomate;devinput;hw_default;avermedia/lircd.conf.vdomate;
+BestBuy Easy TV (BT848) (card=55);bestbuy;devinput;hw_default;bestbuy/lircd.conf.bestbuy;
+BestBuy Easy TV (BT878) (card=62);bestbuy2;devinput;hw_default;bestbuy/lircd.conf.bestbuy2;
+Chronos Video Shuttle II (card=35);chronos;devinput;hw_default;chronos/lircd.conf.chronos;
 Creative BreakOut-Box;breakoutbox;lirc_dev lirc_i2c;hw_default;creative/lircd.conf.breakoutbox;
-Dynalink Magic TView (card=48);cph03x;lirc_dev lirc_gpio;hw_default;cph03x/lircd.conf.cph03x;
-FlyVideo II (card=8);flyvideo;lirc_dev lirc_gpio;hw_default;life-view/lircd.conf.flyvideo;
-FlyVideo 98 (card=30);flyvideo;lirc_dev lirc_gpio;hw_default;life-view/lircd.conf.flyvideo;
-FlyVideo 98/FM /2000S (card=56);flyvideo;lirc_dev lirc_gpio;hw_default;life-view/lircd.conf.flyvideo;
-Flyvideo 98FM (LR50Q) / Typhoon TView TV/FM Tuner (card=36);flyvideo;lirc_dev lirc_gpio;hw_default;life-view/lircd.conf.flyvideo;
+Dynalink Magic TView (card=48);cph03x;devinput;hw_default;cph03x/lircd.conf.cph03x;
+FlyVideo II (card=8);flyvideo;devinput;hw_default;life-view/lircd.conf.flyvideo;
+FlyVideo 98 (card=30);flyvideo;devinput;hw_default;life-view/lircd.conf.flyvideo;
+FlyVideo 98/FM /2000S (card=56);flyvideo;devinput;hw_default;life-view/lircd.conf.flyvideo;
+Flyvideo 98FM (LR50Q) / Typhoon TView TV/FM Tuner (card=36);flyvideo;devinput;hw_default;life-view/lircd.conf.flyvideo;
 Hauppauge TV card;hauppauge;lirc_dev lirc_i2c;hw_default;hauppauge/lircd.conf.hauppauge;
 Hauppauge HVR-1300;hauppauge;lirc_dev lirc_i2c;hw_default;hauppauge/lircd.conf.hauppauge;
 Hauppauge DVB-s card (ver. 2.1);hauppauge_dvb;lirc_dev;hw_default;hauppauge/lircd.conf.hauppauge;
 Hercules Smart TV Stereo (card=100);hercules_smarttv_stereo;lirc_dev lirc_i2c;hw_default;hercules/lircd.conf.smarttv_stereo;
-I-O Data Co. GV-BCTV5/PCI (card=81);gvbctv5pci;lirc_dev lirc_gpio;hw_default;iodata/lircd.conf.gvbctv5pci;
-Jetway TV/Capture JW-TV878-FBK, Kworld KW-TV878RF (card=78);kworld;lirc_dev lirc_gpio;hw_default;kworld/lircd.conf.kworld;
+I-O Data Co. GV-BCTV5/PCI (card=81);gvbctv5pci;devinput;hw_default;iodata/lircd.conf.gvbctv5pci;
+Jetway TV/Capture JW-TV878-FBK, Kworld KW-TV878RF (card=78);kworld;devinput;hw_default;kworld/lircd.conf.kworld;
 KNC ONE TV Station (-/SE/PRO/RDS);knc_one;lirc_dev lirc_i2c;hw_default;knc_one/lircd.conf.knc_one;
-Lenco MXTV-9578 CP (card=50);pixelview_pak;lirc_dev lirc_gpio;hw_default;pixelview/lircd.conf.playtv_pro;
+Lenco MXTV-9578 CP (card=50);pixelview_pak;devinput;hw_default;pixelview/lircd.conf.playtv_pro;
 Miro PCTV serial port receiver;pctv;none;hw_pinsys;pinnacle_systems/lircd.conf.pctv;
-Phoebe Tv Master + FM (card=22);cph06x;lirc_dev lirc_gpio;hw_default;cph03x/lircd.conf.cph03x;
+Phoebe Tv Master + FM (card=22);cph06x;devinput;hw_default;cph03x/lircd.conf.cph03x;
 Pinnacle Systems PCTV Sat receiver;pctv;none;hw_pinsys;pinnacle_systems/lircd.conf.pctv;
-Pixelview PlayTV MPEG2;pixelview_pro;lirc_dev lirc_gpio;hw_default;pixelview/lircd.conf.playtv_pro;
-PixelView PlayTV PAK (card=50);pixelview_pak;lirc_dev lirc_gpio;hw_default;pixelview/lircd.conf.playtv_pro;
-Pixelview PlayTV pro (card=37);pixelview_pro;lirc_dev lirc_gpio;hw_default;pixelview/lircd.conf.playtv_pro;
-Prolink Pixelview PV-BT878P+ (Rev.4C,8E, card=70);pixelview_pro;lirc_dev lirc_gpio;hw_default;pixelview/lircd.conf.playtv_pro;
-Prolink PV-BT878P+4E (card=50);pixelview_pak;lirc_dev lirc_gpio;hw_default;pixelview/lircd.conf.playtv_pro;
+Pixelview PlayTV MPEG2;pixelview_pro;devinput;hw_default;pixelview/lircd.conf.playtv_pro;
+PixelView PlayTV PAK (card=50);pixelview_pak;devinput;hw_default;pixelview/lircd.conf.playtv_pro;
+Pixelview PlayTV pro (card=37);pixelview_pro;devinput;hw_default;pixelview/lircd.conf.playtv_pro;
+Prolink Pixelview PV-BT878P+ (Rev.4C,8E, card=70);pixelview_pro;devinput;hw_default;pixelview/lircd.conf.playtv_pro;
+Prolink PV-BT878P+4E (card=50);pixelview_pak;devinput;hw_default;pixelview/lircd.conf.playtv_pro;
 ProVideo PV951 (card=42);provideo;lirc_dev lirc_i2c;hw_default;provideo/lircd.conf.pv951;
 Technisat MediaFocus I;mediafocusI;none;hw_default;technisat/lircd.conf.mediafocusI;
 Tekram M230 Mach64 (and others bt829 based);tekram_bt829;lirc_dev lirc_bt829;hw_default;tekram/lircd.conf.m230;
-TriTan Technology TView95 CPH03x (card=1);cph03x;lirc_dev lirc_gpio;hw_default;cph03x/lircd.conf.cph03x;
-TView99 CPH063 (card=38);cph06x;lirc_dev lirc_gpio;hw_default;cph03x/lircd.conf.cph03x;
+TriTan Technology TView95 CPH03x (card=1);cph03x;devinput;hw_default;cph03x/lircd.conf.cph03x;
+TView99 CPH063 (card=38);cph06x;devinput;hw_default;cph03x/lircd.conf.cph03x;
 Typhoon TView RDS / FM Stereo (card=53);knc_one;lirc_dev lirc_i2c;hw_default;knc_one/lircd.conf.knc_one;
 Winfast PVR2000 (Linux kernel >=2.6.11 required);leadtek_pvr2000;lirc_dev lirc_i2c;hw_default;leadtek/lircd.conf.PVR2000;
-Winfast TV2000/XP (card=34);leadtek_0010;lirc_dev lirc_gpio;hw_default;leadtek/lircd.conf.RM-0010;
-WinView 601 (card=17);leadtek_0007;lirc_dev lirc_gpio;hw_default;leadtek/lircd.conf.RM-0007;
+Winfast TV2000/XP (card=34);leadtek_0010;devinput;hw_default;leadtek/lircd.conf.RM-0010;
+WinView 601 (card=17);leadtek_0007;devinput;hw_default;leadtek/lircd.conf.RM-0007;
 
 [IrDA hardware]
 SIR IrDA (built-in IR ports);sir;lirc_dev lirc_sir;hw_default;;
diff -urNad lirc-0.8.3~/drivers/lirc_gpio/Makefile.am lirc-0.8.3/drivers/lirc_gpio/Makefile.am
--- lirc-0.8.3~/drivers/lirc_gpio/Makefile.am	2008-05-10 16:29:28.000000000 -0500
+++ lirc-0.8.3/drivers/lirc_gpio/Makefile.am	2008-05-10 16:32:32.000000000 -0500
@@ -5,9 +5,9 @@
 ## this is so that Automake includes the C compiling definitions, and
 ## includes the source files in the distribution.
 EXTRA_PROGRAMS = automake_dummy
-automake_dummy_SOURCES = lirc_gpio.c
+automake_dummy_SOURCES = lirc_gpio.c bttv_deprecated.o
 
 ## there is no *just* object file support in automake.  This is close enough
-module_DATA = lirc_gpio.o
+module_DATA = lirc_gpio.o bttv_deprecated.o
 
 include ../Makefile.common
diff -urNad lirc-0.8.3~/drivers/lirc_gpio/bttv_deprecated.c lirc-0.8.3/drivers/lirc_gpio/bttv_deprecated.c
--- lirc-0.8.3~/drivers/lirc_gpio/bttv_deprecated.c	1969-12-31 18:00:00.000000000 -0600
+++ lirc-0.8.3/drivers/lirc_gpio/bttv_deprecated.c	2008-05-10 16:32:32.000000000 -0500
@@ -0,0 +1,44 @@
+//Temporarily added until a resolution is reached upstream
+//For kernel 2.6.22
+
+#include <linux/version.h>
+#if LINUX_VERSION_CODE == KERNEL_VERSION(2,6,22)
+#include "extra_2.6.22/bttv.h"
+#include "extra_2.6.22/bttvp.h"
+#else
+#include "../drivers/media/video/bt8xx/bttv.h"
+#include "../drivers/media/video/bt8xx/bttvp.h"
+#endif
+
+struct bttv bttvs[BTTV_MAX];
+unsigned int bttv_debug;
+unsigned int bttv_num; /* number of Bt848s in use */
+
+int bttv_get_cardinfo(unsigned int card, int *type, unsigned *cardid)
+{
+    printk("The bttv_* interface is obsolete and will go away,\n"
+           "please use the new, sysfs based interface instead.\n");
+    if (card >= bttv_num) {
+        return -1;
+    }
+    *type = bttvs[card].c.type;
+    *cardid = bttvs[card].cardid;
+    return 0;
+}
+
+wait_queue_head_t* bttv_get_gpio_queue(unsigned int card)
+{
+    struct bttv *btv;
+
+    if (card >= bttv_num) {
+        return NULL;
+    }
+
+    btv = &bttvs[card];
+    if (bttvs[card].shutdown) {
+        return NULL;
+    }
+    return &btv->gpioq;
+}
+
+
diff -urNad lirc-0.8.3~/drivers/lirc_gpio/bttv_deprecated.h lirc-0.8.3/drivers/lirc_gpio/bttv_deprecated.h
--- lirc-0.8.3~/drivers/lirc_gpio/bttv_deprecated.h	1969-12-31 18:00:00.000000000 -0600
+++ lirc-0.8.3/drivers/lirc_gpio/bttv_deprecated.h	2008-05-10 16:32:32.000000000 -0500
@@ -0,0 +1,6 @@
+//Temporarily re-added for Ubuntu Kernel 2.6.22 until upstream
+//Resolution is reached.
+extern wait_queue_head_t* bttv_get_gpio_queue(unsigned int card);
+extern int bttv_get_cardinfo(unsigned int card, int *type,
+                 unsigned int *cardid);
+
diff -urNad lirc-0.8.3~/drivers/lirc_gpio/lirc_gpio.c lirc-0.8.3/drivers/lirc_gpio/lirc_gpio.c
--- lirc-0.8.3~/drivers/lirc_gpio/lirc_gpio.c	2008-05-10 16:32:32.000000000 -0500
+++ lirc-0.8.3/drivers/lirc_gpio/lirc_gpio.c	2008-05-10 16:32:32.000000000 -0500
@@ -63,6 +63,7 @@
 #elif LINUX_VERSION_CODE == KERNEL_VERSION(2,6,22)
 #include "extra_2.6.22/bttv.h"
 #include "extra_2.6.22/bttvp.h"
+#include "bttv_deprecated.h"
 #endif
 
 #if BTTV_VERSION_CODE < KERNEL_VERSION(0, 7, 45)
